akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -0,0 +1,305 @@
+  case object ShuttingDown extends PoolState {
+    override def order: Int = 1
+  }
+
+  // does not accept new tasks, does not process tasks in queue
+  case object ShutDown extends PoolState {
+    override def order: Int = 2
+  }
+
+  // all threads have been stopped, does not process tasks and does not accept new ones
+  case object Terminated extends PoolState {
+    override def order: Int = 3
+  }
+}
+
+final class BoundedTaskQueue(capacity: Int) extends AbstractBoundedNodeQueue[Runnable](capacity)
+
+class AffinityPoolConfigurator(config: Config, prerequisites: DispatcherPrerequisites) extends ExecutorServiceConfigurator(config, prerequisites) {
+
+  sealed trait CPUAffinityStrategy {
+    def javaStrat: AffinityStrategies
+  }
+
+  case object NoAffinityStrategy extends CPUAffinityStrategy {
+    override val javaStrat: AffinityStrategies = AffinityStrategies.ANY
+  }
+  case object SameCoreStrategy extends CPUAffinityStrategy {
+    override val javaStrat: AffinityStrategies = AffinityStrategies.SAME_CORE
+  }
+  case object SameSocketStrategy extends CPUAffinityStrategy {
+    override val javaStrat: AffinityStrategies = AffinityStrategies.SAME_SOCKET
+  }
+  case object DifferentCoreStrategy extends CPUAffinityStrategy {
+    override val javaStrat: AffinityStrategies = AffinityStrategies.DIFFERENT_CORE
+  }
+  case object DifferentSocketStrategy extends CPUAffinityStrategy {
+    override val javaStrat: AffinityStrategies = AffinityStrategies.DIFFERENT_SOCKET
+  }
+
+  private def toAffinityStrategy(s: String): CPUAffinityStrategy = s match {
+    case "any"              ⇒ NoAffinityStrategy
+    case "same-core"        ⇒ SameCoreStrategy
+    case "same-socket"      ⇒ SameSocketStrategy
+    case "different-core"   ⇒ DifferentCoreStrategy
+    case "different-socket" ⇒ DifferentSocketStrategy
+    case x                  ⇒ throw new IllegalArgumentException("[%s] is not a valid cpu affinity strategy [any, same-core, same-socket, different-core, different-socket]!" format x)
+
+  }
+
+  private def toWaitingStrategy(s: String): WaitingStrategy = s match {
+    case "sleep"     ⇒ sleepWaitingStrategy
+    case "yield"     ⇒ yieldWaitingStrategy
+    case "busy-spin" ⇒ busySpinWaitingStrategy
+    case x           ⇒ throw new IllegalArgumentException("[%s] is not a valid waiting strategy[sleep, yield, same-socket, busy-spin]!" format x)
+
+  }
+
+  private val poolSize = ThreadPoolConfig.scaledPoolSize(
+    config.getInt("parallelism-min"),
+    config.getDouble("parallelism-factor"),
+    config.getInt("parallelism-max"))
+  private val affinityGroupSize = config.getInt("affinity-group-size")
+  private val strategies = config.getStringList("cpu-affinity-strategies").map(toAffinityStrategy)
+  private val waitingStrat = toWaitingStrategy(config.getString("worker-waiting-strategy"))
+  private val tf = new AffinityThreadFactory("affinity-thread-fact", strategies.map(_.javaStrat): _*)
```

Comment:
@patriknw 

I feel your concern regarding documentation. It is a bit confusing to me as well. I took the initiative of contacting Peter Lawrey directly in order to get some clarity on the different strategies and this is what our discussion looks like: 

> 
> **Zahari**: Hi there Peter! First of all, thanks for providing this great library which exposes the functionality for thread affinity. I have a simple question as I am a bit confused about the way affinity locks are used. 
> 
> It is especially confusing to me in the context of an AffinityThreadFactory. I am currently working on a custom executor service to be integrated into Akka and what I would like to achieve is the following... say we want to create an executor service of 8 threads. Currently I am using your AffinityThreadFactory that is provided in the library. Assuming I want to pin each of these 8 threads to e separate core and ensure that every time this thread runs will run on the same core it used to run before (in other words each thread has its own reserved core) how would I do that. When creating a ThreadFactory should i provide an affinity strategy of DIFFERENT_CORE or SAME_CORE ? 
> 
> Furthermore, I find this statement a bit hard to understand. Does that mean that anytime we provide an affinity strategy list, we need to make sure that we have ANY at the tail of the list or otherwise it will be ANY by default. 
> 
> **Peter:** Hello, I would use DIFFERENT_CORE with ANY, otherwise unless it can find a core to assign it won't bind to one in particular.
> 
> **Zahari:** I see. So would that effectively mean that in theory if I am to create 16 threads and have 8 cores, it would bind the first 8 to separate  cores and rest of the threads to ANY core ? 
> 
> **Peter:** If you have 8 cores, you only have at most 7 free (the os and everything else needs to run somewhere) Affinity works best if you bind to isolated cpus which are not blocking in which case it makes sense to have less threads than cpus.
> Without binding, the OS scheduler is free to move the threads around. With binding and isolation, you can ensure a specific thread runs on a specific cpu and nothing else. This can dramatically reduce jitter. If you add busy waiting you can ensure the thread never gives up the cpus and minimise latencies.

As far as I take it from the above conversation, it is DIFFERENT_CORE strat should be used along with the thread factory to attempt to disperse the threads among the available cores